File: app/src/main/java/com/isp/app/ui/programs/ProgramsFragment.java
package com.isp.app.ui.programs;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.isp.app.databinding.FragmentProgramsBinding;
import com.isp.app.ui.programs.adapter.ProgramsAdapter;

public class ProgramsFragment extends Fragment {
    private FragmentProgramsBinding binding;
    private ProgramsViewModel viewModel;
    private ProgramsAdapter programsAdapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentProgramsBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(ProgramsViewModel.class);
        programsAdapter = new ProgramsAdapter(program -> {
            NavHostFragment.findNavController(ProgramsFragment.this)
                    .navigate(ProgramsFragmentDirections.actionProgramsToCourses(program.getId()));
        });

        setupRecyclerView();
        observeViewModel();
        viewModel.loadPrograms();
    }

    private void setupRecyclerView() {
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.recyclerView.setAdapter(programsAdapter);
    }

    private void observeViewModel() {
        viewModel.getPrograms().observe(getViewLifecycleOwner(), programs -> programsAdapter.submitList(programs));
        viewModel.getLoading().observe(getViewLifecycleOwner(), isLoading -> binding.progressBar.setVisibility(isLoading ? View.VISIBLE : View.GONE));
        viewModel.getError().observe(getViewLifecycleOwner(), error -> {
            if (error != null) {
                // Show error message
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
} 

File: app/src/main/java/com/college/app/ui/programs/ProgramsFragment.kt
package com.college.app.ui.programs

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.college.app.databinding.FragmentProgramsBinding
import com.college.app.ui.programs.adapter.ProgramsAdapter

class ProgramsFragment : Fragment() {
    private var _binding: FragmentProgramsBinding? = null
    private val binding get() = _binding!!
    private val viewModel: ProgramsViewModel by viewModels()
    private val programsAdapter = ProgramsAdapter { program ->
        findNavController().navigate(
            ProgramsFragmentDirections.actionProgramsToCourses(program.id)
        )
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentProgramsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
        observeViewModel()
        viewModel.loadPrograms()
    }

    private fun setupRecyclerView() {
        binding.recyclerView.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = programsAdapter
        }
    }

    private fun observeViewModel() {
        viewModel.programs.observe(viewLifecycleOwner) { programs ->
            programsAdapter.submitList(programs)
        }

        viewModel.loading.observe(viewLifecycleOwner) { isLoading ->
            binding.progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
        }

        viewModel.error.observe(viewLifecycleOwner) { error ->
            error?.let {
                // Show error message
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
} 